### ISSUE TITLE
[REVIEW]: AutoFunc: A Python package for automating and verifying functional modeling

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2362]
**Submitting author:** @AlexMikes (<a href="http://orcid.org/0000-0002-7873-5758">Alex Mikes</a>)
**Repository:** <a href="https://github.com/AlexMikes/AutoFunc" target ="_blank">https://github.com/AlexMikes/AutoFunc</a>
**Version:** v0.3.1
**Editor:** @gkthiruvathukal
**Reviewers:** @cmccomb, @e-dub, @srmnitc
**Archive:** <a href="https://doi.org/10.5281/zenodo.4482795" target="_blank">10.5281/zenodo.4482795</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/14f9eb408a9cb91c28332fcac1bce505/status.svg)](https://joss.theoj.org/papers/14f9eb408a9cb91c28332fcac1bce505)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/14f9eb408a9cb91c28332fcac1bce505"><img src="https://joss.theoj.org/papers/14f9eb408a9cb91c28332fcac1bce505/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/14f9eb408a9cb91c28332fcac1bce505/status.svg)](https://joss.theoj.org/papers/14f9eb408a9cb91c28332fcac1bce505)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/AlexMikes/AutoFunc" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@cmccomb & @e-dub, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @gkthiruvathukal know.

✨ **Please try and complete your review in the next six weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/AlexMikes/AutoFunc">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@AlexMikes) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/AlexMikes/AutoFunc">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@AlexMikes) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/AlexMikes/AutoFunc">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@AlexMikes) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @cmccomb, @e-dub it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1115/1.1625400 is OK
- 10.1115/1.1289637 is OK
- 10.5281/zenodo.3694300 is OK

MISSING DOIs

- https://doi.org/10.1007/s00163-001-0008-3 may be missing for title: A functional basis for engineering design: Reconciling and evolving previous efforts

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02362/joss.02362/10.21105.joss.02362.pdf)

**Comment by cmccomb** [comment]
@AlexMikes will you please enable issues in your repository?

**Comment by AlexMikes** [comment]
@cmccomb Thanks for being a reviewer and taking the time to give great feedback! I just enabled issues, should I reply here or after you've put your notes in the issue? 

I'll work on adding documentation for the example files and community guidelines in the README.

**Comment by cmccomb** [comment]
> @AlexMikes will you please enable issues in your repository? Once you've done so I'll put my review notes in as issues. For the time being here they are:
> 
> 1. Contribution and authorship: four authors are listed on the work, but only one author has made commits in the software. How did the other authors contribute to this software?
> 2. Example Usage and Functionality Documentation: There are a number of (really cool!) example scripts, but the functionality in these is largely not demonstrated or discussed in the README. Could you add a list of example scripts as a way of highlighting and documenting functionality?
> 3. Community guidelines: no guidance on how to report issues on problems with the software, or to seek support.

@AlexMikes its my pleasure! Really cool piece of software. I've added them here:
1. https://github.com/AlexMikes/AutoFunc/issues/1
2. https://github.com/AlexMikes/AutoFunc/issues/2
3. https://github.com/AlexMikes/AutoFunc/issues/3

**Comment by AlexMikes** [comment]
Thank you @cmccomb for the excellent suggestions. I have completed these and responded to the issues individually.

**Comment by cmccomb** [comment]
> Thank you @cmccomb for the excellent suggestions. I have completed these and responded to the issues individually.

Thanks @AlexMikes! My review is complete, and I recommend accept, @gkthiruvathukal .

**Comment by gkthiruvathukal** [comment]
@cmccomb, This is great news. 

@e-dub, Will you be able to complete your evaluation soon?

**Comment by e-dub** [comment]
> @cmccomb, This is great news.
> 
> @e-dub, Will you be able to complete your evaluation soon?

I would like to start my review but am unable to edit the review checklist. 
@gkthiruvathukal: could you check this for me?

**Comment by gkthiruvathukal** [comment]
@openjournals/joss-eics @e-dub is having trouble to edit the review checklist (and shows as a  proper reviewer). I think we've seen this happen in the past but don't recall exactly how we resolved it.

**Comment by gkthiruvathukal** [comment]
@e-dub You can still start working on identifying issues and opening issues in the submissions project repo and linking those issues here.

**Comment by kthyng** [comment]
@whedon re-invite @e-dub  as reviewer

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@e-dub  please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by kthyng** [comment]
@e-dub See if this helps, maybe you didn't accept the invitation to review previously.

**Comment by gkthiruvathukal** [comment]
@kthyng Thanks for helping to sort this out!

@e-dub Can you let us know when you are able to complete your review?

**Comment by gkthiruvathukal** [comment]
@e-dub Just a gentle nudge to find out when we can expect your review? I know this is a busy time of year for many, including me.

**Comment by gkthiruvathukal** [comment]
@e-dub ?

**Comment by kthyng** [comment]
@gkthiruvathukal Unless you can track down an email address for @e-dub to try another contact venue, this may be a good time to find a replacement reviewer.

**Comment by AlexMikes** [comment]
Thanks, @gkthiruvathukal and @kthyng 

Let me know if I can do anything to help move this along

**Comment by gkthiruvathukal** [comment]
@kthyng Yes, I'll try to get someone else before the weekend. I've been dealing with a rather intense phase at work. I almost went into "out of office" mode as I thought I could work through this.

@AlexMikes  What could help is to suggest more reviewers. I think I've tried all of your initial suggestions. Let me know if you have any others in mind from our reviewer list.

**Comment by AlexMikes** [comment]
I suggest @trallard as an additional reviewer

**Comment by gkthiruvathukal** [comment]
Hi @trallard, Are you willing to contribute a review for this JOSS submission?

**Comment by kthyng** [comment]
@gkthiruvathukal yeah sorry :( Don't want to bug people but yet here I am doing so!

@AlexMikes more suggestions == better :)

**Comment by trallard** [comment]
Hi @gkthiruvathukal at the moment I do not have the bandwidth to contribute to this. I have other reviews that I need to move forward.

**Comment by gkthiruvathukal** [comment]
Thanks, @trallard. I know the feeling. The work just isn't stopping now.

@AlexMikes if you can suggest a couple more, it would help me right now. Thanks!

**Comment by AlexMikes** [comment]
I suggest @moorepants and @galessiorob as potential reviewers

**Comment by arfon** [comment]
@gkthiruvathukal if at all possible we should avoid asking JOSS editors to review (@trallard and @galessiorob) as they're already busy with other JOSS work.

**Comment by moorepants** [comment]
I'm currently reviewing another JOSS paper. Once that is done I can likely take on future ones. So I decline.

**Comment by gkthiruvathukal** [comment]
@arfon I agree 100%. I'll try the other suggestion.

**Comment by gkthiruvathukal** [comment]
It looks like @galessiorob is also a member of joss-editors. So we might need a few more suggestions, @AlexMikes. We'll get there soon, I hope!

**Comment by AlexMikes** [comment]
Ok, how about @nnadeau or @srmnitc?

**Comment by gkthiruvathukal** [comment]
@nnadeau and @srmnitc, are either of you willing to serve as a reviewer for this JOSS submission? Would really appreciate your help!

**Comment by nnadeau** [comment]
@gkthiruvathukal unfortunately, I do not have time right now :(

**Comment by srmnitc** [comment]
@gkthiruvathukal @AlexMikes this module looks interesting and I could find some time to help with the review. However, before I start I would like to actually comment on a major issue. It is great that the software has already been used in publications, however, I do not find any documentation except what is on the README file. I really find this very limiting and would be one blocking issue that I would raise!

**Comment by AlexMikes** [comment]
Thanks, @srmnitc! 

I agree that documentation is important. Every module has NumPy formatted docstrings that explain the functionality and all functions have docstrings that explain the functionality as well as the parameters with their expected type and purpose in the function and the returns with their type and explanation. What else would you suggest for documentation?

**Comment by srmnitc** [comment]
@AlexMikes Yes, the docstrings do look very well written! I was thinking more about using [readthedocs](https://readthedocs.org/) or [github pages](https://pages.github.com/) to host the documentation. Readthedocs then would build the docs including docstrings. It would also allow you to have more details about installing, tests, examples, and so on. Although not required, you could also have interactive examples which can be set up using [mybinder](https://mybinder.org/). I would be happy to raise some issues in detail if I am the reviewer.

**Comment by gkthiruvathukal** [comment]
@srmnitc I am interpreting your last comment as being willing to serve as a reviewer. Can I add you so we can get started?

**Comment by srmnitc** [comment]
@gkthiruvathukal Yes, sorry about the confusion. I am happy to be reviewer.

**Comment by gkthiruvathukal** [comment]
@whedon add @srmnitc as reviewer

**Comment by whedon** [comment]
OK, @srmnitc is now a reviewer

**Comment by gkthiruvathukal** [comment]
Thanks, @srmnitc. You should be all set.

**Comment by srmnitc** [comment]
@gkthiruvathukal Sorry to bother, but I do not seem to have a checklist. Could you please help with this?

**Comment by AlexMikes** [comment]
> @AlexMikes Yes, the docstrings do look very well written! I was thinking more about using [readthedocs](https://readthedocs.org/) or [github pages](https://pages.github.com/) to host the documentation. Readthedocs then would build the docs including docstrings. It would also allow you to have more details about installing, tests, examples, and so on. Although not required, you could also have interactive examples which can be set up using [mybinder](https://mybinder.org/). I would be happy to raise some issues in detail if I am the reviewer.

Thanks @srmnitc, yes please open this as an issue along with anything else you find.

**Comment by danielskatz** [comment]
> @gkthiruvathukal Sorry to bother, but I do not seem to have a checklist. Could you please help with this?

added

**Comment by srmnitc** [comment]
@gkthiruvathukal I have finished my first round of review and opened some issues in the repository. At the moment, it is not possible for me to check many of the boxes, but that should be possible once the issues are cleared. Overall, it seems like a well-written and useful module. 

However, I have to raise some concerns regarding the scholarly effort. The module itself is fairly small, with about 1500 lines or so including the example scripts and test scripts. But the author mentions in the accompanying paper that the module has been used in publications already, which is of course wonderful and warrants publication. The DOIs for these publications are missing at the moment, once those are available I would be happy to go through them.

**Comment by AlexMikes** [comment]
@gkthiruvathukal When I went to look at the issues from @srmnitc, I found some issues from @e-dub from September 18th. So even though he does not seem to be responding here, he still opened some issues. Has he been removed as a reviewer for a lack of communication? Most of the issues he opened are reflected in the issues from @srmnitc, can they be removed since he is not a reviewer?

**Comment by gkthiruvathukal** [comment]
@AlexMikes Yes, but I'd make one amendment to this.  In software engineering, we generally don't delete issues. We simply mark them as resolved. What you could do is make a comment for each of these issues. Since they are duplicates (or near duplicates), it should be easy to link to the other issue in the comments field and mark it closed.  It is my understanding we don't usually remove a reviewer, especially if they did some work toward the review but perhaps experienced challenges that prevented them from finishing it. With COVID-19, many have found themselves unable to finish work. It is even taking me a lot longer to bring submissions to proper closure, so I also want to thank you for being so great during the entire process!

Once you've worked through all the issues and the two active reviewers sign off, we'll be able to move toward acceptance.

**Comment by AlexMikes** [comment]
Thanks, @gkthiruvathukal. I commented on each issue, linked them to a similar issue from @srmnitc (if applicable), and closed them. They are linked to your previous post if you want to check them out.

**Comment by AlexMikes** [comment]
Thanks, @srmnitc for your thoughtful and thorough review! You had great suggestions and insight that helped improve the quality of this software package. I made comments on each issue that you opened. There is one specific issue ([Missing DOIs](https://github.com/AlexMikes/AutoFunc/issues/11)) that I would like your feedback on how you suggest I proceed. The rest of them I hope to have addressed sufficiently, but please comment where appropriate if you suggest further changes.

**Comment by srmnitc** [comment]
@AlexMikes Thanks for your work on the issues. I have been able to check some, but others still remain unchecked. The reasons have been updated in the corresponding issues.

@gkthiruvathukal Thanks for the chance to review this submission. As I have mentioned in a previous comment, there are still two main issues that concern me.

- I am unable to verify the scholarly effort in this work as the mentioned references are without DOIs ([see here](https://github.com/AlexMikes/AutoFunc/issues/11)). From the perspective of the code, it is rather a short module, as such I still remain unable to review the scholarly effort. 

- Although I am able to check some of the review boxes, the overall quality of documentation, tests are not comparable to the recent JOSS publications. This lack of documentation, in my opinion, severely limits the potential usage. The author has also mentioned that there will be no further [work in this direction](https://github.com/AlexMikes/AutoFunc/issues/10). As such no further improvements or maintenance to the code is immediately visible.

Because of these reasons, I cannot recommend the acceptance in JOSS at the moment.

**Comment by gkthiruvathukal** [comment]
@srmnitc I have followed up on the DOI issue in the relevant thread. This appear to be a resolvable issue.

@AlexMikes  Can you comment on the documentation and testing issue and how you plan to address it?

**Comment by gkthiruvathukal** [comment]
@AlexMikes Just a nudge for the status on the documentation/testing issues.

**Comment by AlexMikes** [comment]
@gkthiruvathukal @srmnitc I updated the README for the [testing issue.](https://github.com/AlexMikes/AutoFunc/issues/10)

For the [missing DOIs](https://github.com/AlexMikes/AutoFunc/issues/11), I no longer have access to the LaTeX files to create a pre-print document for submission without the copyright of the conference proceedings where they are published. I can remove them from the references or can wait until the conference proceedings are published and they have DOIs.

For the [documentation issue](https://github.com/AlexMikes/AutoFunc/issues/18), each module has thorough docstrings in standard NumPy format, but I am unable to get readthedocs to autogenerate API documentation for the docstrings. I was unaware this was a requirement in order to submit a paper to JOSS.

For context, this software was used for my graduate school research. I tried submitting this paper in March but submissions were closed due to the pandemic. I graduated in June right around when the submissions finally reopened, and have been working in the industry since then. I spent a lot of time developing this software over previous years and am unable to spend any more time on it. I submitted to JOSS with the hope that others would find it useful in their research. Previous reviewers have found it useful and sufficiently documented, however, if it is the opinion of other reviewers and the editor that it is not sufficient as-is, then please withdraw my submission to JOSS and I apologize for wasting your time.

**Comment by gkthiruvathukal** [comment]
@AlexMikes I am NOT in the camp that would like to see it withdrawn. As the handling editor, I feel like you've been responsive to my & reviewers' feedback. Please don't consider your submission a waste of time. It is not. We don't ever make any author feel unwelcome at JOSS.

Let me summarize my view:

- I am comfortable with your proposed resolution for the missing DOIs. I think, for now, it is best to remove those DOIs, if there is insufficient time to work out the issues. If you have a bit more time, another option might be to post the authors' accepted manuscript to a preprint service such as ArXiv or to ask your co-authors to post to their institutions' digital library (which would give you a DOI as well).

- It is NOT a requirement that readthedocs can autogenerate the docs. If your own Sphinx build works, that is sufficient from my point of view. (I personally host my docs using travis-sphinx and GitHub Pages (and not readthedocs). 

I want to move toward acceptance on this submission and am confident that you have been responsive to feedback. I'm also confident that our remaining reviewers did a great job and are just trying to be helpful. That said, as an editor, I will make the final recommendation, especially since this is well within my areas of expertise. 

Thanks for your response!

**Comment by AlexMikes** [comment]
Thanks, @gkthiruvathukal. 

I checked again and the proceedings were just published for 2/3 of the papers without DOIs. I removed the third that still doesn't have one (the conference is this week).

I got sphinx-autoapi to make some API documentation, I will work on getting some more info on the hosted documentation index.

**Comment by AlexMikes** [comment]
@gkthiruvathukal @srmnitc I addressed the three open issues with resolutions that I hope will be sufficient for final acceptance.

**Comment by gkthiruvathukal** [comment]
@AlexMikes Thanks so much!

@srmnitc Can you take one final look so I can move this forward? I think @AlexMikes has worked to address your feedback.

**Comment by gkthiruvathukal** [comment]
@srmnitc I believe all of your issues have been addressed. Assuming there are no objections, I am going to move forward with acceptance of this submission by end of week. We've kept @AlexMikes waiting a bit longer than I had hoped but with the holidays, I hope it is understandable.

**Comment by gkthiruvathukal** [comment]
@AlexMikes I am ready to move toward acceptance. Please do the following:

- [ ] Make a tagged release of your software, and list the version tag of the archived version here. 
- [ ] Archive the reviewed software in Zenodo 
- [ ] Check the Zenodo deposit has the correct metadata, this includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it); you may also add the authors' ORCID.
- [ ] List the Zenodo DOI of the archived version here.

**Comment by AlexMikes** [comment]
@gkthiruvathukal this is great news, thank you for all of your work in this process!



- [x]  Make a tagged release of your software, and list the version tag of the archived version here.
 **v0.3.1**
- [x]  Archive the reviewed software in Zenodo
- [x]  Check the Zenodo deposit has the correct metadata, this includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it); you may also add the authors' ORCID.
- [x]  List the Zenodo DOI of the archived version here.
**10.5281/zenodo.4482795**

**Comment by gkthiruvathukal** [comment]
@whedon set 10.5281/zenodo.4482795 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4482795" target="_blank">10.5281/zenodo.4482795</a> is the archive.

**Comment by gkthiruvathukal** [comment]
@whedon set v0.3.1 as version

**Comment by whedon** [comment]
OK. v0.3.1 is the version.

**Comment by gkthiruvathukal** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02362/joss.02362/10.21105.joss.02362.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02362/joss.02362/10.21105.joss.02362.pdf) :page_facing_up: :point_left:

**Comment by gkthiruvathukal** [comment]
@AlexMikes My pleasure. *Thank you, too!* for all of your work and patience with us during COVID-19. I'm moving to acceptance now!

**Comment by gkthiruvathukal** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1115/1.1625400 is OK
- 10.1115/1.1289637 is OK
- 10.1007/s00163-001-0008-3 is OK
- 10.5281/zenodo.3694300 is OK

MISSING DOIs

- None

INVALID DOIs

- https://doi.org/10.1115/DETC2020-22495 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1115/DETC2020-22346 is INVALID because of 'https://doi.org/' prefix
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2066

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2066, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by kyleniemeyer** [comment]
Hi @AlexMikes, I'm doing some final checks before publishing - there are just a few small changes needed in the paper:

- We require a brief Statement of Need section that illustrates the research purpose of the software (part of our [required paper elements](https://joss.readthedocs.io/en/latest/submitting.html#what-should-my-paper-contain), normally checked for automatically by whedon). You can likely reuse some of the text in your summary, but add this section header.
- in your reference for the software, it looks like your name is missing a space (the `author` field should be `Alex Mikes` instead of `AlexMikes`)
- in the two 2020 ASME IDETC references (`edmonds2020_2` and `mikes2020`), you need to remove the `https://doi.org/` from the DOI fields

Please let me know when you've made these changes, and then I will publish the paper. There's no need to re-deposit into Zenodo, since we archive the paper itself. Thanks!

**Comment by AlexMikes** [comment]
Thanks, @kyleniemeyer, I have made these changes. Let me know if you need anything else from me.

**Comment by kyleniemeyer** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02362/joss.02362/10.21105.joss.02362.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02362/joss.02362/10.21105.joss.02362.pdf) :page_facing_up: :point_left:

**Comment by kyleniemeyer** [comment]
👍 @AlexMikes, looks good.

**Comment by kyleniemeyer** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1357133759368339456) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2072
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02362](https://doi.org/10.21105/joss.02362)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by kyleniemeyer** [comment]
Congrats @AlexMikes on your article's publication in JOSS!

Many thanks to @cmccomb and @srmnitc for reviewing this submission, and @gkthiruvathukal for editing it.

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02362/status.svg)](https://doi.org/10.21105/joss.02362)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02362">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02362/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02362/status.svg
   :target: https://doi.org/10.21105/joss.02362
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02362/status.svg)](https://doi.org/10.21105/joss.02362)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by AlexMikes** [comment]
Thanks @kyleniemeyer! Thanks again @gkthiruvathukal for leading this and @cmccomb and @srmnitc for reviewing and helping to improve it!

**Comment by xyluo25** [comment]
Hi @AlexMikes, I hope you’re doing well and enjoying a productive week. I recently submitted the pyufunc paper to JOSS (just yesterday), and I was delighted to see that JOSS recommended you as an ideal reviewer given your expertise. I know how busy schedules can be, so I would be incredibly grateful if you could kindly consider reviewing the manuscript. Your insights and feedback would mean a great deal to me and help ensure pyufunc meets the highest standards.

Please let me know if you have any questions or need additional information. Thank you so much for your time and consideration. I truly appreciate it.
